proxy_worker/dispatcher.py

```diff
@@ -0,0 +1,528 @@
+                request_handler = self._handle__invocation_request
+            case _:
+                # Don't crash on unknown messages. Log the error and return.
+                logger.error("Unknown StreamingMessage content type: %s", content_type)
+                return
+
+        resp = await request_handler(request)
+        self._grpc_resp_queue.put_nowait(resp)
+
+    async def dispatch_forever(self):  # sourcery skip: swap-if-expression
+        if DispatcherMeta.__current_dispatcher__ is not None:
+            raise RuntimeError('there can be only one running dispatcher per '
+                               'process')
+
+        self._old_task_factory = self._loop.get_task_factory()
+
+        DispatcherMeta.__current_dispatcher__ = self
+        try:
+            forever = self._loop.create_future()
+
+            self._grpc_resp_queue.put_nowait(
+                protos.StreamingMessage(
+                    request_id=self.request_id,
+                    start_stream=protos.StartStream(
+                        worker_id=self.worker_id)))
+
+            # In Python 3.11+, constructing a task has an optional context
+            # parameter. Allow for this param to be passed to ContextEnabledTask
+            self._loop.set_task_factory(
+                lambda loop, coro, context=None: ContextEnabledTask(
+                    coro, loop=loop, context=context))
+
+            # Detach console logging before enabling GRPC channel logging
+            logger.info('Detaching console logging.')
+            disable_console_logging()
+
+            # Attach gRPC logging to the root logger. Since gRPC channel is
+            # established, should use it for system and user logs
+            logging_handler = AsyncLoggingHandler()
+            root_logger = logging.getLogger()
+
+            log_level = logging.INFO if not is_envvar_true(
+                PYTHON_ENABLE_DEBUG_LOGGING) else logging.DEBUG
+
+            root_logger.setLevel(log_level)
+            root_logger.addHandler(logging_handler)
+            logger.info('Switched to gRPC logging.')
+            logging_handler.flush()
+
+            try:
+                await forever
+            finally:
+                logger.warning('Detaching gRPC logging due to exception.')
+                logging_handler.flush()
+                root_logger.removeHandler(logging_handler)
+
+                # Reenable console logging when there's an exception
+                enable_console_logging()
+                logger.warning('Switched to console logging due to exception.')
+        finally:
+            DispatcherMeta.__current_dispatcher__ = None
+
+            self._loop.set_task_factory(self._old_task_factory)
+            self.stop()
+
+    def stop(self) -> None:
+        if self._grpc_thread is not None:
+            self._grpc_resp_queue.put_nowait(self._GRPC_STOP_RESPONSE)
+            self._grpc_thread.join()
+            self._grpc_thread = None
+
```

Comment:
[Blocker] You are joining thread and cleanup, there could be exception from these which will left resources in open(or undefined state).  I will encourage use to exception and finally to ensure thread safety.

Reply:
>which will left resources in open(or undefined state)

This is already at the end of the workers lifetime - and we can discard any state.